src/imagedef/types/gif.ts

```diff
@@ -23,30 +24,64 @@ import ImageDefinition from '../../imagedef';
  *          1 bytes  <Trailer> (0x3b)
  * *
  */
-export default (buffer): ImageDefinition => {
+export default async (aBuffer: AsyncBuffer): Promise<ImageDefinition> => {
   const BLOCK_TERMINATOR = 0x00;
   const EXTENSION_INTRODUCER = 0x21;
   const GRAPHIC_CONTROL_LABEL = 0xf9;
   const TRANSPARENT_COLOR_FLAG = 0x01;
   const IMAGE_SEPARATOR = 0x2c;
   const INTERLACE_FLAG = 0x64;
-  if (buffer.length < 13 ||
+
+  await aBuffer.waitForSize(13);
+  if (aBuffer.buffer.length < 13 ||
       // "GIF" 0x47 0x49 0x46
-      buffer[0] !== 0x47 || buffer[1] !== 0x49 || buffer[2] !== 0x46 ||
+      aBuffer.buffer[0] !== 0x47 || aBuffer.buffer[1] !== 0x49 || aBuffer.buffer[2] !== 0x46 ||
       // "87a" or "89a"
-      buffer[3] !== 0x38 || (buffer[4] !== 0x39 && buffer[4] !== 0x37) || buffer[5] !== 0x61) {
+      aBuffer.buffer[3] !== 0x38 ||
+      (aBuffer.buffer[4] !== 0x39 && aBuffer.buffer[4] !== 0x37) || aBuffer.buffer[5] !== 0x61) {
     return;
   }
-  const width = buffer.readUInt16LE(6);
-  const height = buffer.readUInt16LE(8);
-  let interlacing = false;
-  let alpha = false;
-  let animated = false;
+  const width = aBuffer.buffer.readUInt16LE(6);
+  const height = aBuffer.buffer.readUInt16LE(8);
+
+  // version 8[7]a
+  // This version cant be animated and does not support alpha
+  // so these values are known, but we do need to find
+  // whether interlacing is used
+  if (aBuffer.buffer[4] === 0x37) {
+    const interlacingOffset = 14 + 8; // 14 = GIF header
+    await aBuffer.waitForSize(interlacingOffset + 1);
+    // tslint:disable-next-line:no-bitwise
+    const interlacing = !(aBuffer.buffer[interlacingOffset] & INTERLACE_FLAG);
+    return {
+      type: 'gif',
+      width,
+      height,
+      interlacing,
+      alpha: false,
+      animated: false,
+    };
+  }
 
   // version: 8[9]a
-  if (buffer[4] === 0x39) {
+  // This version can both be animated and supports alpha
+  // It consists of either:
+  // Graphic Control Extension + Image Block
+  // or:
+  // Application Extension
+  // [ Graphic Control Extension + Image Block ]* (each frame)
+
+  if (aBuffer.buffer[4] === 0x39) {
     // graphic control extension.
+
+    // Havent figured out how to parse GIF in a nice way without
+    // reading full image yet.
+    await aBuffer.waitUntilEnd();
```

Comment:
Can't you just in the end of the loop wait until the next byte? We're just looking for the block terminator. Or another method is to add a new method waitForByte(BLOCK_TERMINATOR).

Reply:
it looks really good though!